source/particle/generator/quadrature_points.cc

```diff
@@ -29,62 +29,85 @@ namespace aspect
     namespace Generator
     {
       template <int dim>
+      QuadraturePoints<dim>::QuadraturePoints()
+      {
+        starting_particle_index = 0;
+      }
+
+      template <int dim>
       void
       QuadraturePoints<dim>::generate_particles(std::multimap<types::LevelInd, Particle<dim> > &particles)
       {
-        types::particle_index particle_index = 0;
-
         const QGauss<dim> quadrature_formula(this->get_parameters().stokes_velocity_degree + 1);
 
+        unsigned int n_particles_to_generate = 0, prefix_sum = 0;
+
         FEValues<dim> fe_values(this->get_mapping(),
                                 this->get_fe(),
                                 quadrature_formula,
                                 update_values |
                                 update_quadrature_points |
                                 update_JxW_values);
 
-        typename Triangulation<dim>::active_cell_iterator
+        typename parallel::distributed::Triangulation<dim>::active_cell_iterator
         cell = this->get_triangulation().begin_active(),
         endc = this->get_triangulation().end();
-        for (; cell != endc; cell++)
+
+        if (cell != endc)
           {
             fe_values.reinit(cell);
-            std::vector<Point<dim>> quadrature_points = fe_values.get_quadrature_points();
-            for (typename std::vector<Point<dim>>::const_iterator q_points_itr = quadrature_points.begin();
-                 q_points_itr != quadrature_points.end(); q_points_itr++)
-              {
-                Point<dim> particle_position_real = (*q_points_itr);
+            n_particles_to_generate = this->get_triangulation().n_locally_owned_active_cells() * fe_values.n_quadrature_points;
+          }
+
+        MPI_Scan(&n_particles_to_generate, &prefix_sum, 1, MPI_UNSIGNED, MPI_SUM, this->get_mpi_communicator());
```

Comment:
Given the number of particles to generate locally, compute the appropriate offset in the global generation of particles and store the result in variable prefix_sum.